code/modules/library/lib_machines.dm

```diff
@@ -213,3 +214,109 @@ var/global/list/library_section_names = list("Fiction", "Non-Fiction", "Adult",
+	icon = 'icons/obj/library.dmi'
+	icon_state = "binder"
+	anchored = TRUE
+	density = TRUE
+	machine_flags =  WRENCHMOVE | FIXED2WORK | EJECTNOTDEL // | SCREWTOGGLE | CROWDESTROY
+	pass_flags = PASSTABLE
+	idle_power_usage = 4
+	var/obj/item/weapon/disk/tech_disk/diskslot
+	var/busy = FALSE
+
+/obj/machinery/researcharchive/New()
+	..()
+	if(!research_archive_datum)
+		research_archive_datum = new /datum/research()
+
+/obj/machinery/researcharchive/examine(mob/user)
+	..()
+	if(diskslot)
+		to_chat(user,"<span class='info'>In the slot you can see a disk that contains [diskslot.stored.id] [diskslot.stored.level].</span>")
+		diskslot.examine(user)
+
+/obj/machinery/researcharchive/attackby(var/obj/item/weapon/W, var/mob/user)
+	if(stat & (BROKEN))
+		to_chat(user, "<span class='warning'>\The [src] is broken!</span>")
+		return
+	if(..())
+		return
+
+	if(busy)
+		return
+
+	if (!istype(W,/obj/item/weapon/disk/tech_disk))
+		to_chat(user, "<span class='warning'>\The [src] only accepts technology disks.</span>")
+		return
+
+	if (!user.drop_item(W, src))
+		return
+	if(diskslot)
+		user.put_in_hands(diskslot)
+		visible_message("<span class='notice'>\The [user] swaps the disks in \the [src].</span>","<span class='notice'>You swap the disks in \the [src].</span>")
+	else
+		visible_message("<span class='notice'>\The [user] adds \the [W] to \the [src].</span>","<span class='notice'>You add \the [W] to \the [src].</span>")
+	diskslot = W
+	playsound(loc, 'sound/machines/click.ogg', 50, 1)
+	update_icon()
+
+/obj/machinery/researcharchive/attack_hand(var/mob/user)
+	. = ..()
+	if(stat & (BROKEN))
+		to_chat(user, "<span class='notice'>\The [src] is broken.</span>")
+		return
+
+	if(stat & (NOPOWER))
+		to_chat(user, "<span class='notice'>\The [src] is unpowered.</span>")
+		return
+
+	if(!diskslot)
+		to_chat(user, "<span class='notice'>There is no inserted technology disk.</span>")
+		return
+
+	if(busy)
+		return
+
+	if(diskslot.stored.id in list("syndicate", "Nanotrasen", "anomaly"))
+		to_chat(user, "<span class='notice'>\The [src] cannot process this technology data due to proprietary encoding.</span>")
+		return
+
+	playsound(loc, "sound/machines/heps.ogg", 50, 1)
+
+	busy = TRUE
+	use_power(200)
+	spawn(3 SECONDS)
+
+	for(var/datum/tech/T in get_list_of_elements(research_archive_datum.known_tech))
+		if(T.id != diskslot.stored.id)
+			continue
+		if(T.level > diskslot.stored.level)
+			if(T.level>=6)
+				visible_message("<span class='warning'>\The [src] rejects the data disk as [T.id] data has already reached its maximum.")
+				break
+			//Pick the lowest: +3 levels, level 6, maximum level of tech, or the maximum level on the disk
+			//Example: increase to level 4 in one pass, then increase to level 6 in second pass
+			T.level = min(T.level+3, 6, T.max_level, diskslot.stored.level)
+			qdel(diskslot)
+			diskslot = null
```

Comment:
You can do `QDEL_NULL(diskslot)`, which is exactly the same thing but in one line